crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3347,19 +3401,45 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 
     fn infer_slice_expression(&mut self, slice: &ast::ExprSlice) -> Type<'db> {
+        enum SliceArg {
+            Arg(Option<i32>),
+            Unsupported,
+        }
+
         let ast::ExprSlice {
             range: _,
             lower,
             upper,
             step,
         } = slice;
 
-        self.infer_optional_expression(lower.as_deref());
-        self.infer_optional_expression(upper.as_deref());
-        self.infer_optional_expression(step.as_deref());
+        let ty_lower = self.infer_optional_expression(lower.as_deref());
+        let ty_upper = self.infer_optional_expression(upper.as_deref());
+        let ty_step = self.infer_optional_expression(step.as_deref());
 
-        // TODO slice
-        Type::Todo
+        let type_to_slice_argument = |ty: Option<Type<'db>>| match ty {
+            Some(Type::IntLiteral(n)) if i32::try_from(n).is_ok() => {
+                SliceArg::Arg(Some(i32::try_from(n).expect("checked in branch arm")))
+            }
```

Comment:
> I guess you could avoid the `.expect()` call here by doing something like:

Yes. I added the other three uses of this before. And there I really need/want it as a pattern guard, because I want control flow to fall through if `i32::try_from` fails. Because I don't want to repeat the whole logic for the "else" clause.

But right here, your solution is definitely better. Changed in #13982